spec/amq/uri_parsing_spec.rb

```diff
@@ -12,7 +16,140 @@
+        context "absent" do
+          let(:uri) { "amqp://" }
+
+          # Note that according to the ABNF, the host component may not be absent, but it may be zero-length.
+          it "fallbacks to default nil host" do
+            expect(subject[:host]).to be_nil
+          end
+        end
+      end
+
+      if RUBY_VERSION < "2.2"
+        context "absent" do
+          let(:uri) { "amqp://" }
+
+          # Note that according to the ABNF, the host component may not be absent, but it may be zero-length.
+          it "raises InvalidURIError" do
+            expect { subject[:host] }.to raise_error(InvalidURIError, /bad URI(absolute but no path)/)
+          end
+        end
+      end
+    end
+
+    describe "port" do
+      context "present" do
+        let(:uri) { "amqp://rabbitmq:5672" }
+
+        it "parses port" do
+          expect(subject[:port]).to eq(5672)
+        end
+      end
+
+      context "absent" do
+        context "schema amqp" do
+          let(:uri) { "amqp://rabbitmq" }
+
+          it "fallbacks to 5672 port" do
+            expect(subject[:port]).to eq(5672)
+          end
+        end
+
+        context "schema amqps" do
+          let(:uri) { "amqps://rabbitmq" }
+
+          it "fallbacks to 5671 port" do
+            expect(subject[:port]).to eq(5671)
+          end
+        end
+      end
+    end
+
+    describe "username and passowrd" do
+      context "both present" do
+        let(:uri) { "amqp://alpha:beta@rabbitmq" }
+
+        it "parses user and pass" do
+          expect(subject[:user]).to eq("alpha")
+          expect(subject[:pass]).to eq("beta")
+        end
+      end
+
+      context "only username present" do
+        let(:uri) { "amqp://alpha@rabbitmq" }
+
+        it "parses user and fallbacks to nil pass" do
+          expect(subject[:user]).to eq("alpha")
+          expect(subject[:pass]).to be_nil
+        end
+
+        context "with ':'" do
+          let(:uri) { "amqp://alpha:@rabbitmq" }
+
+          it "parses user and fallbacks to "" (empty) pass" do
+            expect(subject[:user]).to eq("alpha")
+            expect(subject[:pass]).to eq("")
+          end
+        end
+      end
+
+      context "only password present" do
+        let(:uri) { "amqp://:beta@rabbitmq" }
+
+        it "parses pass and fallbacks to "" (empty) user" do
+          expect(subject[:user]).to eq("")
+          expect(subject[:pass]).to eq("beta")
+        end
+      end
+
+      context "both absent" do
+        let(:uri) { "amqp://rabbitmq" }
+
+        it "fallbacks to nil user and pass" do
+          expect(subject[:user]).to be_nil
+          expect(subject[:pass]).to be_nil
+        end
+
+        context "with ':'" do
+          let(:uri) { "amqp://:@rabbitmq" }
+
+          it "fallbacks to "" (empty) user and "" (empty) pass" do
+            expect(subject[:user]).to eq("")
+            expect(subject[:pass]).to eq("")
+          end
+        end
+      end
+
+      context "escaped" do
```

Comment:
I believe you mean "%-encoded".

Reply:
According to method names in [URI::Escape](https://ruby-doc.org/stdlib-2.2.2/libdoc/uri/rdoc/URI/Escape.html) module of `URI` library, [escaped](https://ruby-doc.org/stdlib-2.2.2/libdoc/uri/rdoc/URI/Escape.html#method-i-escape) and [encoded](https://ruby-doc.org/stdlib-2.2.2/libdoc/uri/rdoc/URI/Escape.html#method-i-encode) mean the same. But your variant looks more descriptive and unambiguous ^_^